infrastructures/infrastructure-common/src/main/java/org/ow2/proactive/resourcemanager/nodesource/infrastructure/util/InitScriptGenerator.java

```diff
@@ -0,0 +1,193 @@
+                                                      nodeSourceName,
+                                                      nodeBaseName,
+                                                      numberOfNodesPerInstance,
+                                                      credentials);
+        // use the unified line breaker "\n"
+        startupScript = startupScript.replace("\r\n", "\n");
+        List<String> startupScriptList = Arrays.stream(startupScript.split("\n"))
+                                               .filter(s -> !s.isEmpty())
+                                               .collect(Collectors.toList());
+
+        logger.info("Linux startup script generated: " + startupScriptList);
+        return startupScriptList;
+    }
+
+    public List<String> buildWindowsScript(String startupScriptTemplate, String instanceId, String rmUrl,
+            String rmHostname, String nodeJarUrl, String instanceIdNodeProperty, String additionalProperties,
+            String nodeSourceName, String nodeBaseName, int numberOfNodesPerInstance, String credentials) {
+        String startupScript = startupScriptTemplate;
+        // the complete parameter part of additionalProperties needs to be removed in windows powershell script when it's empty
+        if (additionalProperties.isEmpty()) {
+            startupScript = startupScript.replace(WINDOWS_COMPLETE_ADDITIONAL_PROPERTIES_PROPERTY, "");
+        }
+        startupScript = fillInScriptProperties(startupScript,
+                                               instanceId,
+                                               rmUrl,
+                                               rmHostname,
+                                               nodeJarUrl,
+                                               instanceIdNodeProperty,
+                                               additionalProperties,
+                                               nodeSourceName,
+                                               nodeBaseName,
+                                               numberOfNodesPerInstance,
+                                               credentials);
+        // use the unified line breaker "\n"
+        startupScript = startupScript.replace("\r\n", "\n");
+        startupScript = Arrays.stream(startupScript.split("\n"))
+                              .map(String::trim)
+                              .filter(s -> !s.isEmpty())
+                              .map(s -> s.endsWith(";") ? s : s + ";")
+                              .collect(Collectors.joining());
+        startupScript = POWERSHELL_COMMAND_PREFIX + startupScript + POWERSHELL_COMMAND_SUFFIX;
+        List<String> startupScriptList = Collections.singletonList(startupScript);
+        logger.info("Windows startup script generated: " + startupScriptList);
+        return startupScriptList;
+    }
+
+    protected String fillInScriptProperties(String startupScriptTemplate, String instanceId, String rmUrl,
```

Comment:
Really nice feature. I was just wondering if the provision script supplied by the user could also benefit from the macro substitution feature you have implemented in this method.

Reply:
Yes, it's supported.